[Chua Zi Long] iP

ChuaZiLong opened the pull request:
**New Features Implementation** ✨ 
This pull request introduces a new feature that allows users to search for tasks based on a keyword.

What's included:

- Introduction of FindCommand class
- Integration of the find feature into the existing Parser class
- Enhancement of the Ui class to display search results

Changes in Detail

1. FindCommand Class: Handles the logic for searching tasks by keyword.
2. Parser Class Updates: Parses user input and recognizes the find command.
3. Ui Class Enhancements: Displays the tasks that match the search criteria.

Sample Usage
```
// Example of command usage
todo book
find book
```

Tasks Completed
 Implement the FindCommand class.
 Update the Parser class to recognize the find command.
 Enhance the Ui class to show search results.
 Write unit tests for the new feature (work in progress).

⚠️  Compatibility concerns: Please note that this feature requires **Java 11** or later.
For more details, check out the [feature documentation](https://docs.oracle.com/javase/8/docs/).

How to use:
Simply type find <keyword> in the command interface, and the system will list all tasks containing the given <keyword>.

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```
public static void main(String[] args) {
        new Duke("./data/duke.txt/duke.txt").run();
}
```

Happy searching! 🔍

Commit:
Level-0

Commit:
no message

Commit:
no message

Commit:
no message

Commit:
no message

Commit:
no message

Commit:
no message

Commit:
no message

Commit:
Save tasks in the hard disk automatically

Commit:
no message

Commit:
no message

Commit:
no message

Commit:
no message

Commit:
Package duke

Commit:
Add Gradle

Commit:
Add Gradle

Commit:
Add Gradle

Commit:
no message

Commit:
Add tests

Commit:
no message

Commit:
Merge branch 'master' of https://github.com/ChuaZiLong/ip
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

Commit:
no message

Commit:
no message

Commit:
Add JavaDocs

Commit:
Add a way to find todos, deadlines and events with an inputted keyword

Commit:
no message

Commit:
no message

Commit:
no message

Commit:
no message

Commit:
no message

Commit:
no message

Commit:
no message

Commit:
Remove unnecessary comment lines from the code

Commit:
Add placeholder comment lines to the code

Commit:
Remove unnecessary comment lines from the code

Commit:
Remove unnecessary comment lines from the code

Commit:
Introduce basic features of JavaFX into the program

Commit:
Introduce JavaFX into the program

Commit:
Integrate JavaFX into the existing commands in the programme

solomonng2001 commented on src/main/java/seedu/duke/Duke.java:
Hi Zi Long, good job overall! :smile:

I noticed that there were many duplicates of Duke.java in various other folders. Would it be neater to remove the other duplicates?

solomonng2001 commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,784 @@
+ * Represents the user interface of the Duke application.
+ * Handles interactions with the user.
+ */
+class Ui {
```
Would it be more organised to abstract the class UI into a separate package UI? I noticed the same for Storage, TaskList, Task, Parser etc, which are all currently in 1 Duke.java file. Organising the file structure might make a big project more maintainable and easier to work on by fellow developers.

solomonng2001 commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,784 @@
+        System.out.println("Noted. I've removed this task:");
+        System.out.println("  " + task);
+        System.out.println("Now you have " + taskCount + " tasks in the list.");
+    }
```
Would it be more readable to add a next line between showDeletedTak and showFindResults?

solomonng2001 commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,784 @@
+        System.out.println("  " + task);
+        System.out.println("Now you have " + taskCount + " tasks in the list.");
+    }
+    public void showFindResults(ArrayList<Task> tasks) {
```
Did you intend to add java doc comments for the showFindResults method?

solomonng2001 commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,784 @@
+     * @return An ArrayList of tasks loaded from the file.
+     * @throws DukeException If the file is not found or the tasks cannot be loaded.
+     */
+    public ArrayList<Task> load() throws DukeException {
```
Would it be more intuitive for future readers if the method was renamed to something more specific like loadTaskFromStorage etc? I noticed the same for the "save" method

solomonng2001 commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,784 @@
+                    Task task = null;
+                    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
+
+                    switch (type) {
```
Would it be more readable if case was indented on the same level as switch?
<img width="1007" alt="Screenshot 2024-02-07 at 6 54 19 PM" src="https://github.com/nus-cs2103-AY2324S2/ip/assets/66112309/9ff52a51-c8e3-473b-9cc8-4fbe79fd984a">

solomonng2001 commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,784 @@
+     * @param task The task to be converted to a string.
+     * @return A formatted string representing the task.
+     */
+    private String taskToFileString(Task task) {
```
Would it be more intuitive to rename the method name to an action like getStorageFormat etc?

solomonng2001 commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,784 @@
+     * @return A formatted string representing the task.
+     */
+    private String taskToFileString(Task task) {
+        String type = task instanceof Todo ? "T" :
```
Would it be more readable if "if-else" statements were used instead of the ternary operator?

solomonng2001 commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,784 @@
+                int index = Integer.parseInt(commandArgs) - 1;
+                return new MarkCommand(index);
+            } catch (NumberFormatException e) {
+                throw new DukeException("Invalid task number format.");
```
Would there be less duplication if a static final variable was used to store the string here? Consider the case that the developer decides to make minor changes to this string, if the string was stored "INVALID_TASK_NUMBER = ...", and all other codes reference that, at least only that constant needs to be edited rather than all other instances where it was used 😃 

I noticed the same for other strings, where there were also duplicates

solomonng2001 commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,784 @@
+
+    public String getStatusIcon() {
+
+        return "[" + taskType + "]" + (isDone ? "[X] " : "[ ] ");
```
Would it be easier to use String.format() like in the following image, instead of having many concatenation operators? Actually, I think this is personally preference, but would like to let you know of alternatives. Concatenation could be more readable too, since variables and string can be read sequentially.

<img width="465" alt="Screenshot 2024-02-07 at 7 07 14 PM" src="https://github.com/nus-cs2103-AY2324S2/ip/assets/66112309/dc661595-16d4-4fba-a795-bb713647afd1">

solomonng2001 commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,784 @@
+    }
+}
+
+abstract class Command {
```
Amazing use of OOP here!!! 👍

solomonng2001 reviewed (commented):
Great job Zi Long 👍 

The quality of the OOP is well designed. I hope my comments will help you improve the minor syntax-related items.

Commit:
Integrate JavaFX into existing methods

Commit:
Merge branch 'Level-10'

Commit:
Add JavaDoc comments to the code

Commit:
Add JavaDoc comments to the code

Commit:
Add JavaDoc comments to the code

Commit:
Add JavaDoc comments to the code

Commit:
Add JavaDoc comments to the code

Commit:
Add JavaDoc comments to the code

Commit:
Add JavaDoc comments to the code

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Tweak the code to comply with the coding standard

Commit:
Tweak the code to comply with the coding standard

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Tweak the code to comply with the coding standard

Commit:
Add JavaDoc comments to the code

Commit:
Add the find command to the code

Commit:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Duke.java

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-Level-9'

modembcc commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,784 @@
+
+class AddDeadlineCommand extends Command {
+    private String description;
+    private LocalDateTime by;
```
Should the variable name be more descriptive as it is used throughout the class? Ex. deadline, target_date, etc.

modembcc commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,784 @@
+}
+
+class Deadline extends Task {
+    protected LocalDateTime by;
```
Should the variable name be more descriptive as it is used throughout the class? Ex. deadline, target_date, etc.

modembcc commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,784 @@
+import java.io.IOException;
+import java.io.FileWriter;
+
+class DukeException extends Exception {
```
Since the bot is no longer Duke should we rename the exception to reflect the name of the chatbot?

modembcc commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,784 @@
+ * Represents the main class for the Duke application.
+ * Initializes the application and starts the interaction with the user.
+ */
+public class Duke {
```
Since the bot is no longer Duke should we rename the class to reflect the name of the chatbot?

modembcc commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,784 @@
+    public static Command parse(String fullCommand) throws DukeException {
+        String[] commandParts = fullCommand.split(" ", 2);
+        String commandType = commandParts[0];
+        String commandArgs = commandParts.length > 1 ? commandParts[1] : "";
```
Should we rename commandArgs as a singular as it is not a collection (yet)? Ex. commandArgLine, argString, etc.

modembcc reviewed (commented):
Overall, I think you did a good job at naming most of the variables. It made the code very easy to understand.

Commit:
Fix buggy code that appeared out of nowhere somehow

Commit:
No assert lines now

No checks for assumptions on program state

Add assert lines to some parts of the code

Gets runtime to check the assumptions

Commit:
Now there are a few long methods

This makes it difficult for others to understand the code

Refactor the code in those parts

Makes the code more readable

Commit:
Egyptian style braces are not properly used in some places

Makes code hard to read

Change the code in these places

Makes the code easier to read

ChuaZiLong committed:
Merge pull request #2 from ChuaZiLong/branch-A-CodeQuality

Improve code quality in some ways

Commit:
Add placeholder comment to the code

Commit:
Remove placeholder comments from the code

Commit:
Merge branch 'master' of https://github.com/ChuaZiLong/ip

Commit:
Improve the find command so that it displays items even if the keyword matches the item only partially

Commit:
Add a User Guide to the project

Commit:
Add more tests for Parser and TaskList

Add more JavaDocs to the code

Commit:
Shorten start method in the Duke class

Commit:
Change variable name to fix a bug

Commit:
Update the screenshot of Ui.png

Commit:
Remove unnecessary comments from the code
